File: clients/python/random_player.py
import asyncio
import json
import random
import sys

import websockets


def parse_csv(value: str):
    value = value.strip()
    if not value:
        return []
    return [v.strip() for v in value.split(",") if v.strip()]

def sort_discard_keys(key: str):
    if key.startswith("player") and key.endswith("_discards"):
        num = key[len("player"):-len("_discards")]
        if num.isdigit():
            return (0, int(num))
    return (1, key)


async def run(uri: str, player_id: str, room_id: str):
    async with websockets.connect(uri) as ws:
        print(f"Python client joining room={room_id} player_id={player_id}", flush=True)
        join_msg = {"type": "join", "room_id": room_id, "player_id": player_id, "role": "player"}
        await ws.send(json.dumps(join_msg))

        while True:
            raw = await ws.recv()
            msg = json.loads(raw)
            mtype = msg.get("type")
            if mtype == "state":
                legal = parse_csv(msg.get("legal", ""))
                if legal:
                    # Requesting discards reveals all players' discard information.
                    discards_req = {"type": "discards_request", "room_id": room_id}
                    await ws.send(json.dumps(discards_req))
                    choice = random.choice(legal)
                    action = {
                        "type": "action",
                        "room_id": room_id,
                        "player_id": player_id,
                        "choice": choice,
                    }
                    await ws.send(json.dumps(action))
            elif mtype == "discards":
                discards = {k: v for k, v in msg.items() if k.endswith("_discards")}
                if discards:
                    parts = [f"{k}={discards[k]}" for k in sorted(discards, key=sort_discard_keys)]
                    # print("discards " + " ".join(parts))
            elif mtype == "game_over":
                winner = msg.get("winner")
                print(f"game_over winner={winner}")
                return
            elif mtype == "error":
                print(f"error: {msg.get('message')}")


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("usage: python random_player.py ws://localhost:9002 room1 p1")
        sys.exit(1)
    uri = sys.argv[1]
    room_id = sys.argv[2] if len(sys.argv) > 2 else "room1"
    player_id = sys.argv[3] if len(sys.argv) > 3 else "p1"
    asyncio.run(run(uri, player_id, room_id))

File: clients/java/RandomPlayer.java
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.LinkedBlockingQueue;

public class RandomPlayer {
    private static class Listener implements WebSocket.Listener {
        private final BlockingQueue<String> messages;
        private final StringBuilder buffer = new StringBuilder();

        Listener(BlockingQueue<String> messages) {
            this.messages = messages;
        }

        @Override
        public CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
            buffer.append(data);
            if (last) {
                messages.add(buffer.toString());
                buffer.setLength(0);
            }
            webSocket.request(1);
            return CompletableFuture.completedFuture(null);
        }
    }

    private static String extractString(String json, String key) {
        String pattern = "\"" + key + "\"";
        int pos = json.indexOf(pattern);
        if (pos < 0) return null;
        pos = json.indexOf(':', pos + pattern.length());
        if (pos < 0) return null;
        pos++;
        while (pos < json.length() && Character.isWhitespace(json.charAt(pos))) pos++;
        if (pos >= json.length() || json.charAt(pos) != '\"') return null;
        pos++;
        int end = json.indexOf('\"', pos);
        if (end < 0) return null;
        return json.substring(pos, end);
    }

    private static Integer extractInt(String json, String key) {
        String pattern = "\"" + key + "\"";
        int pos = json.indexOf(pattern);
        if (pos < 0) return null;
        pos = json.indexOf(':', pos + pattern.length());
        if (pos < 0) return null;
        pos++;
        while (pos < json.length() && Character.isWhitespace(json.charAt(pos))) pos++;
        int end = pos;
        while (end < json.length() && (Character.isDigit(json.charAt(end)) || json.charAt(end) == '-')) end++;
        if (end == pos) return null;
        return Integer.parseInt(json.substring(pos, end));
    }
    private static List<String> parseCsv(String value) {
        List<String> items = new ArrayList<>();
        if (value == null || value.trim().isEmpty()) return items;
        for (String token : value.split(",")) {
            String t = token.trim();
            if (!t.isEmpty()) items.add(t);
        }
        return items;
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.out.println("usage: java RandomPlayer ws://localhost:9002 room1 p1");
            return;
        }
        String uri = args[0];
        String roomId = args.length > 1 ? args[1] : "room1";
        String playerId = args.length > 2 ? args[2] : "p1";

        BlockingQueue<String> messages = new LinkedBlockingQueue<>();
        HttpClient client = HttpClient.newHttpClient();
        WebSocket webSocket = client.newWebSocketBuilder()
                .buildAsync(URI.create(uri), new Listener(messages))
                .join();

        System.out.println("Java client joining room=" + roomId + " player_id=" + playerId);
        String join = String.format("{\"type\":\"join\",\"room_id\":\"%s\",\"player_id\":\"%s\",\"role\":\"player\"}",
                roomId, playerId);
        webSocket.sendText(join, true).join();

        Random rng = new Random();
        while (true) {
            String msg = messages.take();
            String type = extractString(msg, "type");
            if (type == null) {
                continue;
            }
            if (type.equals("state")) {
                String legal = extractString(msg, "legal");
                List<String> choices = parseCsv(legal);
                if (!choices.isEmpty()) {
                    // Requesting discards reveals all players' discard information.
                    String discardsReq = String.format("{\"type\":\"discards_request\",\"room_id\":\"%s\"}", roomId);
                    webSocket.sendText(discardsReq, true).join();
                    String choice = choices.get(rng.nextInt(choices.size()));
                    String action = String.format("{\"type\":\"action\",\"room_id\":\"%s\",\"player_id\":\"%s\",\"choice\":\"%s\"}",
                            roomId, playerId, choice);
                    webSocket.sendText(action, true).join();
                }
            } else if (type.equals("discards")) {
                // System.out.println("discards " + msg);
            } else if (type.equals("game_over")) {
                Integer winner = extractInt(msg, "winner");
                System.out.println("game_over winner=" + winner);
                break;
            } else if (type.equals("error")) {
                String message = extractString(msg, "message");
                System.out.println("error: " + message);
            }
        }

        webSocket.sendClose(WebSocket.NORMAL_CLOSURE, "bye").join();
    }
}
